lib/linked_list.rb

```diff
@@ -12,120 +11,264 @@ def initialize(value, next_node = nil)
-    def add_last(value)
-      raise NotImplementedError
+  end
+  
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def visit    
+    current = @head
+    
+    until current.nil?
+      print current.data
+      current = current.next
     end
-
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    def get_last
-      raise NotImplementedError
+  end
+  
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def delete(value)
+    current = @head
+    previous = nil
+    
+    until current.nil?
+      if current.data == value
+        if previous.nil?
+          @head = current.next
+          return current.next = current.next.next
+        else
+          previous.next = current.next
+          break
+        end 
+      end 
+      
+      previous = current
+      current = current.next
     end
-
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    def insert_ascending(value)
-      raise NotImplementedError
+  end 
+  
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n)
+  # Space Complexity: O(1) - reverse in place
+  def reverse
+    return if @head.nil?
+    
+    current = @head
+    previous = nil
+    
+    until current.next.nil?
+      temp = current.next
+      current.next = previous
+      previous = current
+      current = temp
+    end 
+    
+    current.next = previous
+    @head = current
+  end
+  
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity:
+  # Space Complexity:
+  def find_middle_value
+    raise NotImplementedError
+  end
+  
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n+m) - two loops with different lengths
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    return if @head.nil?
+    current = @head
+    length = 0
+    
+    until current.next.nil?
+      length += 1
+      current = current.next
+    end 
+    
+    return nil if n > length
+    
+    current = @head
+    (length - n).times do |i|
+      current = current.next
+    end 
+    
+    return current.data
+  end
+  
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity:
+  # Space Complexity:
+  def has_cycle
+    raise NotImplementedError
+  end
+  
+  # Additional Exercises 
+  # returns the value in the first node
+  # returns nil if the list is empty
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_first
+    return if @head.nil?

```

Comment:
This works, but consider

```suggestion
    return nil if @head.nil?

```